doc/source/specification.rst

```diff
@@ -1,11 +1,96 @@
 Specification
 =============
 Description of how MorphIO sees morphologies. MorphIO supports :ref:`.asc <specification-neurolucida>` (Neurolucida),
-`.h5`_ (H5 version 1) and `.swc`_ extensions.
+`.h5`_ (H5 version 1) and `.swc`_ extensions. A morphology is represented as a soma with neurites. Either soma or
+neurites can be absent. Neurite is a tree of sections. Section is composed of segments where segment is made of points.
+Lets start describing them from the smallest to the biggest entity.
 
-Soma formats:
--------------
-MorphIO implements Soma specification from `NeuroMorpho`_, and recognizes several kinds of soma format.
+
+Point
+-----
+A point is an array of three numbers **[X, Y, Z]** where X, Y, Z are Cartesian coordinates of position.
+A point always has a diameter associated with it.
+
+
+Segment
+-------
+A segment consists of two consecutive points(`Point`_) belonging to the same section(`Section`_).
+
+.. image:: images/swc/two_pt.svg
+   :scale: 100 %
+   :alt: segment
+
+
+Section
+-------
+
+A section is a series of one or more segments. Each section has a type associated with it. The type shows what part of
+the neuron a section represents. The type can of the following: axon, soma, basal dendrite, apical dendrite or
+custom dendrite. `Section API <cpp/doxygen_Section.html>`_.
+
+.. image:: images/section_segment.svg
+   :scale: 100 %
+   :alt: section
+
+The first and the last point of section must be of the following combinations:
+
+1. starting point, forking point
+2. forking point, end point
+3. forking point, forking point
+4. starting point, end point
+
+Starting point is the first point of a section without parent. Forking point is the last point of a section that has
+multiple children. End point is the last point of a section without children. For more details on section's parent and
+children see `Neurite`_. The above combinations are marked with their corresponding number at the image below.
+
+.. image:: images/sections.svg
+   :scale: 100 %
+   :alt: section variants
+
+Neurite
+-------
+
+A neurite is a tree of sections(`Section`_). The tree structure implies the following:
+
+* Section can have only one parent (another section).
+* Section can have an arbitrary number of children (other sections).
+* No loops are present in the structure.
+
+A section without parent is called a root section. A section with parent must have its first point to be a duplicate
+of the last point of its parent.
+
+Section ordering
+****************
+In MorphIO each section is identified by an ID. By default, the section IDs will correspond to
+the order of section appearance while performing a depth-first traversal on every neurites. The
```

Comment:
I am not sure we a have a stability guaranty with DFS. Isn't the DFS a consequence of the file formats (specially for swc or ascii) ?

Reply:
What is DFS? Depth-first-...? I agree that this statement is wrong. I bluntly copied it from NeuroM. I need to edit it.

Reply:
Depth-first-search sorry

Reply:
Because it depends on the ordering of the neighbors and so it depends on the graph / tree underlying implementation. If you are using something like a adjacency sets (with unordered set) you cannot guaranty a stable iterator on the tree/graph.

Reply:
Can you please help me with clarification on this.  I'm lost in the sources of MorphIO.

Reply:
Yes I am checking

Reply:
so, I'm putting this into docs? Would it be ok for you?

Reply:
lol actually after thinking about it and having a better understanding, I think the doc is correct...

Reply:
XD, should I put something like "verified by Tom" among the lines?

Reply:
"TomDele seal of approval".